src/utils/promise-pool.js

```diff
@@ -0,0 +1,96 @@
+function getRandomString() {
+    return Math.random().toString(36).substring(2, 15);
+}
+
+export default class PromisePool {
+    constructor() {
+        /**
+         * @type {Object.<string, Promise>}
+         */
+        this.items = {};
+        /**
+         * @type {Array.<Function>}
+         */
+        this.subscriptionCallbacks = [];
+        /**
+         * @type {Array.<string>}
+         */
+        this.settledItems = [];
```

Comment:
When a promise gets settled (either resolved or rejected), it goes into this list here